README.md
# worldtides

API wrapper for [WorldTides.info](https://www.worldtides.info) on [Node.js](https://nodejs.org) (unofficial)

[![npm](https://img.shields.io/npm/v/worldtides.svg?maxAge=3600)](https://github.com/fvdm/nodejs-worldtides/blob/master/CHANGELOG.md)
[![Build Status](https://github.com/fvdm/nodejs-worldtides/actions/workflows/node.js.yml/badge.svg?branch=master)](https://github.com/fvdm/nodejs-worldtides/actions/workflows/node.js.yml)
[![Coverage Status](https://coveralls.io/repos/github/fvdm/nodejs-worldtides/badge.svg?branch=master)](https://coveralls.io/github/fvdm/nodejs-worldtides?branch=master)


## Usage

You provide the parameters in an _object_ and it returns a _Promise_.

When a parameter only needs to be 'present' you still need to give it a value,
like an empty string `''` or `true`, it can be anything.

The `date`, `localtime` and `start` params are auto formatted from strings and Date.

See the [**API documentation**](https://www.worldtides.info/apidocs) for details.


```js
const tides = require ('worldtides');

// Get tide extremes for the next 2 weeks (3 credits)
tides ({
  key: 'abc123',
  extremes: '',
  length: 1209600,
  stationDistance: 15,
  lat: 52.381158,
  lon: 4.4837275,
})
.then (data => data.extremes)
.then (console.table)
.catch (console.error);
```


## Installation

`npm i worldtides`


## Configuration

You can add the configuration params in the same arguments
object as the API request method.

```js
const tides = require ('worldtides');

tides ({
  key: 'abc123',
  timeout: 5000,
  ...params,
});
```

### Parameters

param     | type   | default     | description
:---------|:-------|:------------|:-----------
[origin]  | string | `no-origin` | Origin on allow list
[timeout] | number | `5000`      | Request time out in ms


## Unlicense

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>


## Author

[Franklin](https://fvdm.com)
| [Buy me a coffee](https://fvdm.com/donating)


CHANGELOG.md
## 2.0.0 (2021-10-06)

##### Breaking Changes

*  Full rewrite ([6b797153](https://github.com/fvdm/nodejs-worldtides/commit/6b797153c0c6de4ef323f1dc706b06f554fcac9f))

##### Chores

* **package:**
  *  Update dependencies ([b23d2c6a](https://github.com/fvdm/nodejs-worldtides/commit/b23d2c6aeacee5f46a149372f461308924d1015b))
  *  Update runkit url ([12e870ce](https://github.com/fvdm/nodejs-worldtides/commit/12e870ceb9d656525ccdbc12c0fffedabf316886))

##### Documentation Changes

*  Nicer examples with console.table ([bce60b94](https://github.com/fvdm/nodejs-worldtides/commit/bce60b94ce9897cc37ec304523ee1106f7814b2b))
*  Fixed config example ([7a85c4dc](https://github.com/fvdm/nodejs-worldtides/commit/7a85c4dc48188d89d7d32508b4664b4ae4825a1d))
*  Note about date params ([7c6715ce](https://github.com/fvdm/nodejs-worldtides/commit/7c6715ce89d74dda146457f5a246c7eeacdf4796))
*  Fixed table, minor clean up ([cb4682e5](https://github.com/fvdm/nodejs-worldtides/commit/cb4682e57afdf42be0756ac7ca71d4f5403b2c65))
*  Removed double example ([c2c5b809](https://github.com/fvdm/nodejs-worldtides/commit/c2c5b8098adeef90812ff7e2b87807d93258b058))
*  Include API key in example ([2e3f4b95](https://github.com/fvdm/nodejs-worldtides/commit/2e3f4b95e1d899401e45c394d3f393bf2b1489e9))
* **license:**  Updated https url ([bde99605](https://github.com/fvdm/nodejs-worldtides/commit/bde996053c52be4a579bd636429e0e645084dab3))

##### New Features

*  Auto convert input dates ([fdeea9d3](https://github.com/fvdm/nodejs-worldtides/commit/fdeea9d37b772a5e15f23532df2bd8c762a9e69e))

##### Bug Fixes

*  Dates fixing ([cb23af1b](https://github.com/fvdm/nodejs-worldtides/commit/cb23af1b979ee01053d3f756dcc2a5fc586d8044))
*  Missing request parameters ([74821bac](https://github.com/fvdm/nodejs-worldtides/commit/74821bac682fe5bcda1a31d9ffc8798b974eb8c8))
*  Syntax typo ([6b014bac](https://github.com/fvdm/nodejs-worldtides/commit/6b014bace3eadf09c15ffc33aed67ab50f01b50f))
*  Lint errors ([e4e31ac9](https://github.com/fvdm/nodejs-worldtides/commit/e4e31ac96da4a8fcc7c3f70ed8e55eaa9d228db8))
*  Syntax typo ([1395ad93](https://github.com/fvdm/nodejs-worldtides/commit/1395ad933a072dbcdcded51062120af6ad58d9b3))

##### Refactors

* **package:**  Minimum node v12 ([3cb17ce8](https://github.com/fvdm/nodejs-worldtides/commit/3cb17ce8ca8b910bd614d03dca80174e3768413f))
*  Moved fixDates() to main function ([296bbd68](https://github.com/fvdm/nodejs-worldtides/commit/296bbd6855f685d8105b0d4d93736a7472394ee6))
*  Clean up parameters ([16cd99b3](https://github.com/fvdm/nodejs-worldtides/commit/16cd99b3c507e88bf19315e7f6baf23b1ba76d0a))
*  Missing key is handled by the API ([b9531fc4](https://github.com/fvdm/nodejs-worldtides/commit/b9531fc40860c47fce322e867ccf83d12f146ae9))
*  Cleaner date conversion ([3c15391a](https://github.com/fvdm/nodejs-worldtides/commit/3c15391a9c7ab5847b023e27f4f149953efe3438))
*  Updated API endpoint ([94c7f40b](https://github.com/fvdm/nodejs-worldtides/commit/94c7f40b833c2f12d84dc825480be4b83a640d38))
*  Parameters ([ff9fa188](https://github.com/fvdm/nodejs-worldtides/commit/ff9fa18827e165e52e3a27b84a93ddb5a9e83432))
*  Another attempt ([54fd7e16](https://github.com/fvdm/nodejs-worldtides/commit/54fd7e160fa04ad2913348c61ae08e9ad6862644))
*  Export syntax ([ce1aad45](https://github.com/fvdm/nodejs-worldtides/commit/ce1aad45d9fcd0a030bde643c37f67e700f34dad))

##### Code Style Changes

*  Added comments ([3a682840](https://github.com/fvdm/nodejs-worldtides/commit/3a682840c0b2b8c315eadd05764f896cbfce1398))

##### Tests

*  Fixed typo ([6540e3bc](https://github.com/fvdm/nodejs-worldtides/commit/6540e3bc1dedbf05349eb3eeadadb013d6aea110))
*  Added date conversion checks ([26bd223f](https://github.com/fvdm/nodejs-worldtides/commit/26bd223f416549952c3ee7664ef747627ce1eecc))
*  Fixed bad ref ([e2aba576](https://github.com/fvdm/nodejs-worldtides/commit/e2aba57619d0c9de2330fe85fa9b5b66a631d635))
*  Refactored combined config ([f155771b](https://github.com/fvdm/nodejs-worldtides/commit/f155771b466b22beb582cf06043be3a23422cbd9))
*  Refactored config assignment ([7e4edb90](https://github.com/fvdm/nodejs-worldtides/commit/7e4edb909b69cfa17f5b0c14b1336c63f12abb73))
*  Fixed syntax error ([e75557ce](https://github.com/fvdm/nodejs-worldtides/commit/e75557ce7706ea3af03258a87037ae672d7b5dbd))
*  Fixed again ([f94a5240](https://github.com/fvdm/nodejs-worldtides/commit/f94a52401cd827c7ea47c3a17765ebe82a09deb8))
*  Fixed ([91d76b01](https://github.com/fvdm/nodejs-worldtides/commit/91d76b017cfe623f98c2e5fe887b6f4ce1dc4f60))
* **config:**  Fixed ES env in ESLint ([d4ea2bd1](https://github.com/fvdm/nodejs-worldtides/commit/d4ea2bd1c08cb40f9166cadecf2e7f4463e6ccd2))
* **ci:**  Removed old git fetch ([de38df7c](https://github.com/fvdm/nodejs-worldtides/commit/de38df7c2830ff4a8972a2ec6b007e3eb394b88a))

#### 1.0.2 (2018-09-02)

##### Chores

* **example:**  correctedConstituents is removed ([2c9aa363](https://github.com/fvdm/nodejs-worldtides/commit/2c9aa3633839305fde0a5712c2b7b58f2107df67))

##### Documentation Changes

* **readme:**  correctedConstituents is removed ([6f7f37fb](https://github.com/fvdm/nodejs-worldtides/commit/6f7f37fb4fe8607058714719f3888c9f30c83451))

##### Refactors

* **package:**  Include example.js in files ([f8cb2774](https://github.com/fvdm/nodejs-worldtides/commit/f8cb277415feb8cf38df3e5503b6c696285bca75))

##### Code Style Changes

* **test:**  Reduced indentation whitespace ([3563c31e](https://github.com/fvdm/nodejs-worldtides/commit/3563c31ef71946ad0907c0229a01e29733f071d2))
* **lint:**  Don't indent MemberExpression ([df1c5818](https://github.com/fvdm/nodejs-worldtides/commit/df1c5818ff73d223af9e1f2bba2102f6ea5fc0f3))
* **talk:**  Less indentation on promises ([33985203](https://github.com/fvdm/nodejs-worldtides/commit/339852032c61569f351a8e507a0b59cea46b2a8c))

##### Tests

* **main:**  Removed apikey from config ([10997e6b](https://github.com/fvdm/nodejs-worldtides/commit/10997e6bcfffb9d72d0df4f276cfb04503ea47c9))


test.js
const dotest = require ('dotest');
const app = require ('./');

const config = {
  key: process.env.KEY || '',
  origin: process.env.ORIGIN || '',
  timeout: process.env.TIMEOUT || '',
};

const params = {
  extremes: '',
  stationDistance: 15,
  lat: 52.381158,
  lon: 4.4837275,
};


dotest.add ('Interface', test => {
  test()
    .isFunction ('fail', 'exports', app)
    .done()
  ;
});


dotest.add ('Communication', async test => {
  let error;
  let data;

  try {
    data = await app ({ ...config, ...params });
  }
  catch (err) {
    error = err;
  }

  test (error)
    .isObject ('fail', 'data', data)
    .isNotEmpty ('fail', 'data', data)
    .done()
  ;
});


dotest.add ('API error', async test => {
  let error;
  let data;

  try {
    data = await app (params);
  }
  catch (err) {
    error = err;
  }

  test()
    .isError ('fail', 'err', error)
    .isNotEmpty ('fail', 'err.message', error && error.message)
    .isUndefined ('fail', 'data', data)
    .done()
  ;
});


dotest.add ('date: today', async test => {
  let error;
  let data;

  try {
    data = await app ({
      ...config,
      ...params,
      date: 'today',
    });
  }
  catch (err) {
    error = err;
  }

  test (error)
    .isObject ('fail', 'data', data)
    .isNotEmpty ('fail', 'data', data)
    .done()
  ;
});


dotest.add ('date: string', async test => {
  let error;
  let data;

  try {
    data = await app ({
      ...config,
      ...params,
      date: '2021-10-01 12:34',
    });
  }
  catch (err) {
    error = err;
  }

  test (error)
    .isObject ('fail', 'data', data)
    .isNotEmpty ('fail', 'data', data)
    .done()
  ;
});


dotest.add ('localtime', async test => {
  let error;
  let data;

  try {
    data = await app ({
      ...config,
      ...params,
      localtime: new Date().toString(),
    });
  }
  catch (err) {
    error = err;
  }

  test (error)
    .isObject ('fail', 'data', data)
    .isNotEmpty ('fail', 'data', data)
    .done()
  ;
});


dotest.add ('start: number', async test => {
  let error;
  let data;

  try {
    data = await app ({
      ...config,
      ...params,
      start: Math.floor (new Date().getTime() / 1000) - 86400,
    });
  }
  catch (err) {
    error = err;
  }

  test (error)
    .isObject ('fail', 'data', data)
    .isNotEmpty ('fail', 'data', data)
    .done()
  ;
});


dotest.add ('start: date', async test => {
  let error;
  let data;

  try {
    data = await app ({
      ...config,
      ...params,
      start: new Date(),
    });
  }
  catch (err) {
    error = err;
  }

  test (error)
    .isObject ('fail', 'data', data)
    .isNotEmpty ('fail', 'data', data)
    .done()
  ;
});


dotest.run ();


index.js
/**
 * Name:               worldtides - index.js
 * Description:        API wrapper for WorldTides.info on Node.js (unofficial)
 * Author:             Franklin <https://fvdm.com>
 * License:            Unlicense (public domain, see LICENSE file)
 * Source & docs:      <https://github.com/fvdm/nodejs-worldtides>
 * API documentation:  <https://www.worldtides.info/apidocs>
 */

const { doRequest } = require ('httpreq');


/**
 * Communicate with the API
 *
 * @param   {object}  params                     Method parameters
 * @param   {string}  [params.origin=no-origin]  CORS origin
 * @param   {number}  [params.timeout=5000]      Request timeout in ms
 *
 * @return  {Promise<object>}
 */

module.exports = async function ({
  origin = 'no-origin',
  timeout = 5000,
}) {
  delete arguments[0].origin;
  delete arguments[0].timeout;

  // Fix dates
  if (arguments[0].date && arguments[0].date !== 'today') {
    arguments[0].date = new Date (arguments[0].date).toISOString().split ('T')[0];
  }

  if (arguments[0].localtime) {
    arguments[0].localtime = new Date (arguments[0].localtime).toISOString();
  }

  if (arguments[0].start && typeof arguments[0].start !== 'number') {
    arguments[0].start = Math.floor (new Date (arguments[0].start).getTime() / 1000);
  }

  // Process request
  const options = {
    url: 'https://www.worldtides.info/api/v2',
    method: 'POST',
    parameters: arguments[0],
    timeout,
    headers: {
      'Origin': origin,
      'User-Agent': 'nodejs-worldtides',
    },
  };

  const res = await doRequest (options);
  const data = JSON.parse (res.body);

  // API error
  if (data.error) {
    throw new Error (data.error);
  }

  // Success
  return data;
};
